File: app/src/main/java/com/smoothradio/radio/feature/discover/ui/adapter/CategoryRecyclerViewAdapter.java
package com.smoothradio.radio.feature.discover.ui.adapter;

import static com.smoothradio.radio.service.StreamService.StreamStates.BUFFERING;
import static com.smoothradio.radio.service.StreamService.StreamStates.PLAYING;
import static com.smoothradio.radio.service.StreamService.StreamStates.PREPARING;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.smoothradio.radio.R;
import com.smoothradio.radio.databinding.CategoryradioitemBinding;
import com.smoothradio.radio.feature.discover.ui.DiscoverFragment;
import com.smoothradio.radio.core.model.RadioStation;
import com.smoothradio.radio.service.StreamService;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class CategoryRecyclerViewAdapter extends RecyclerView.Adapter {
    private final List<RadioStation> favouriteStations = new ArrayList<>();
    private final List<RadioStation> radioStationItems;
    private final DiscoverFragment.RadioStationActionHandler radioStationActionHandler;
    private final DiscoverRecyclerViewAdapter discoverRecyclerViewAdapter;
    private String state = StreamService.StreamStates.ENDED;
    private RadioStation lastStation;
    private boolean isPlaying;



    CategoryRecyclerViewAdapter(DiscoverRecyclerViewAdapter discoverRecyclerViewAdapter, List<RadioStation> stationLogosList, DiscoverFragment.RadioStationActionHandler radioStationActionHandler) {
        this.discoverRecyclerViewAdapter = discoverRecyclerViewAdapter;
        radioStationItems = new ArrayList<>(stationLogosList);
        this.radioStationActionHandler = radioStationActionHandler;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        CategoryradioitemBinding categoryradioitemBinding = CategoryradioitemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ItemViewViewHolder(categoryradioitemBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        RadioStation radioStation = radioStationItems.get(position);

        ItemViewViewHolder itemViewViewHolder = (ItemViewViewHolder) holder;

        updateFavoriteIcon(itemViewViewHolder, radioStation);

        itemViewViewHolder.binding.ivCategoryLogo.setImageResource(radioStation.getLogoResource());
        itemViewViewHolder.binding.tvCategoryChannelName.setText(radioStation.getStationName());
        itemViewViewHolder.binding.ivCategoryPlay.setOnClickListener(new PlayOnclickListener(radioStation));
        itemViewViewHolder.binding.ivCategoryFavourite.setOnClickListener(new FavouriteListener(radioStation, itemViewViewHolder));

        updatePlayer(radioStation, itemViewViewHolder);

        //scroll animation, set it only when the view is newly bound
        if (holder.getBindingAdapterPosition() == RecyclerView.NO_POSITION) {
            itemViewViewHolder.binding.clCategoryLayout
                    .setAnimation(AnimationUtils.loadAnimation(holder.itemView.getContext(), R.anim.recyclerviewscrollanimation));
        }
    }

    private void updatePlayer(RadioStation radioStation, ItemViewViewHolder holder) {
        boolean isPlaying = radioStation.isPlaying();
        if (!isPlaying) {
            holder.binding.lottieLoadingAnimation.setVisibility(View.INVISIBLE);
            holder.binding.ivCategoryPlay.setVisibility(View.VISIBLE);
            holder.binding.ivCategoryPlay.setImageResource(R.drawable.playicon);
            return;
        }

        if (Objects.equals(state, PLAYING)) {
            holder.binding.lottieLoadingAnimation.setVisibility(View.INVISIBLE);
            holder.binding.ivCategoryPlay.setVisibility(View.VISIBLE);
            holder.binding.ivCategoryPlay.setImageResource(R.drawable.pauseicon);
        } else if (Objects.equals(state, PREPARING) || Objects.equals(state, BUFFERING)) {
            holder.binding.lottieLoadingAnimation.setVisibility(View.VISIBLE);
            holder.binding.ivCategoryPlay.setVisibility(View.INVISIBLE);
        } else {
            holder.binding.lottieLoadingAnimation.setVisibility(View.INVISIBLE);
            holder.binding.ivCategoryPlay.setVisibility(View.VISIBLE);
            holder.binding.ivCategoryPlay.setImageResource(R.drawable.playicon);
        }
    }


    @Override
    public int getItemCount() {
        return radioStationItems.size();
    }


    private class ItemViewViewHolder extends RecyclerView.ViewHolder {
        CategoryradioitemBinding binding;

        public ItemViewViewHolder(@NonNull CategoryradioitemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    private class PlayOnclickListener implements View.OnClickListener {
        private final RadioStation radioStation;

        PlayOnclickListener(RadioStation radioStation) {
            this.radioStation = radioStation;
        }

        @Override
        public void onClick(View view) {
            if (lastStation != null) {
                int lastIndex = getPositionOfStation();
                if (lastIndex != RecyclerView.NO_POSITION) {
                    radioStationItems.get(lastIndex).setPlaying(false);
                    notifyItemChanged(lastIndex);
                }
            }

            // Start play
            radioStationActionHandler.onStationSelected(radioStation);

            // Register this adapter as the currently playing one
            discoverRecyclerViewAdapter.setPlayingCategoryAdapter(CategoryRecyclerViewAdapter.this);
        }
    }


    private class FavouriteListener implements View.OnClickListener {
        private final RadioStation radioStation;
        private final ItemViewViewHolder viewHolder;

        FavouriteListener(RadioStation radioStation, ItemViewViewHolder viewHolder) {
            this.radioStation = radioStation;
            this.viewHolder = viewHolder;
        }

        @Override
        public void onClick(View view) {
            boolean isFavorite = favouriteStations.contains(radioStation);

            if (isFavorite) {
                radioStationActionHandler.onRemoveFromFavorites(radioStation.getStationName());
                radioStationActionHandler.onRequestshowToast("Removed from favorites: " + radioStation.getStationName());
            } else {
                if (favouriteStations.size() >= 20) {
                    radioStationActionHandler.onRequestshowToast("Can't add more than 20 favorite stations");
                    return;
                }
                radioStationActionHandler.onAddToFavorites(radioStation.getStationName());
                radioStationActionHandler.onRequestshowToast("Added to favorites: " + radioStation.getStationName());
            }

            // Update UI icon immediately
            updateFavoriteIcon(viewHolder, radioStation);
        }
    }

    private void updateFavoriteIcon(ItemViewViewHolder holder, RadioStation station) {
        int resId = favouriteStations.contains(station)
                ? R.drawable.favorite_20px_filled
                : R.drawable.favorite_20px;
        holder.binding.ivCategoryFavourite.setImageResource(resId);
    }

    public void setFavouriteStations(List<String> favouriteListNames) {
        List<RadioStation> favoriteStations = getFavoriteStationsFromNames(radioStationItems, favouriteListNames);
        favouriteStations.clear();
        favouriteStations.addAll(favoriteStations);
    }

    private List<RadioStation> getFavoriteStationsFromNames(List<RadioStation> allStations, List<String> favoriteNames) {
        List<RadioStation> favoriteStations = new ArrayList<>();
        for (RadioStation station : allStations) {
            if (favoriteNames.contains(station.getStationName())) {
                favoriteStations.add(station);
            }
        }
        return favoriteStations;
    }

    public void setState(String state) {
        this.state = state;
    }

    public void setSelectedStation(RadioStation radioStation) {
        this.lastStation = radioStation;
    }

    public void updateCategory() {
        if (Objects.equals(state, PLAYING)) {
            isPlaying = true;
        } else if (Objects.equals(state, PREPARING) || Objects.equals(state, BUFFERING)) {
            //do nothing
        } else {
            isPlaying = false;
        }
    }

    public void updateStation(RadioStation radioStation) {
        radioStation.setPlaying(radioStation.isPlaying());
        int index = radioStationItems.indexOf(radioStation);
        if (index != RecyclerView.NO_POSITION) {
            notifyItemChanged(index);
        }

    }

    private int getPositionOfStation() {
        return radioStationItems.indexOf(lastStation);
    }

    public boolean isPlaying() {
        return isPlaying;
    }

    public interface RadioStationActionListener {
        void onStationSelected(RadioStation station);

        void onAddToFavorites(String stationName);

        void onRemoveFromFavorites(String stationName);

        void onRequestshowToast(String message);
    }
}

File: app/src/main/kotlin/com/smoothradio/radio/feature/discover/ui/adapter/CategoryRecyclerViewAdapter.kt
package com.smoothradio.radio.feature.discover.ui.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import androidx.core.view.isInvisible
import androidx.core.view.isVisible
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.smoothradio.radio.R
import com.smoothradio.radio.core.domain.model.RadioStation
import com.smoothradio.radio.databinding.CategoryradioitemBinding
import com.smoothradio.radio.feature.discover.util.RadioStationActionHandler
import com.smoothradio.radio.feature.radio_list.ui.adapter.util.StationDiffUtilCallback
import com.smoothradio.radio.service.StreamService
import com.smoothradio.radio.service.StreamService.StreamStates.BUFFERING
import com.smoothradio.radio.service.StreamService.StreamStates.PLAYING
import com.smoothradio.radio.service.StreamService.StreamStates.PREPARING

/**
 * Adapter for displaying a list of radio stations in a nested RecyclerView.
 * This adapter handles the display of station logos, names, play/pause buttons,
 * and favorite icons. It also manages the state of the currently playing station
 * and updates the UI accordingly.
 *
 * @property stationLogosList The initial list of [RadioStation] objects to display.
 * @property radioStationActionHandler An instance of [RadioStationActionHandler] to
 *                                    handle user interactions with the radio stations,
 *                                    such as selecting a station or toggling its favorite status.
 */
class CategoryRecyclerViewAdapter(
    radioStationList: List<RadioStation>,
    private val radioStationActionHandler: RadioStationActionHandler
) : RecyclerView.Adapter<CategoryRecyclerViewAdapter.ItemViewViewHolder>() {

    private val radioStationItems: MutableList<RadioStation> = ArrayList(radioStationList)
    private var state: String = StreamService.StreamStates.ENDED

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewViewHolder {
        val binding =
            CategoryradioitemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ItemViewViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ItemViewViewHolder, position: Int) {
        val radioStation = radioStationItems[position]
        val binding = holder.binding

        updateFavoriteIcon(radioStation, holder)
        updatePlayer(radioStation, holder)

        with(binding) {
            ivCategoryLogo.setImageResource(radioStation.logoResource)
            tvCategoryChannelName.text = radioStation.stationName
            ivCategoryPlay.setOnClickListener(PlayOnclickListener(radioStation))
            ivCategoryFavourite.setOnClickListener(FavouriteListener(radioStation))
        }

        if (holder.adapterPosition == RecyclerView.NO_POSITION) {
            holder.binding.clCategoryLayout.animation =
                AnimationUtils.loadAnimation(
                    holder.itemView.context,
                    R.anim.recyclerviewscrollanimation
                )
        }
    }

    private fun updatePlayer(radioStation: RadioStation, holder: ItemViewViewHolder) =
        with(holder.binding) {
            if (!radioStation.isPlaying) {
                lottieLoadingAnimation.isInvisible = true
                ivCategoryPlay.isVisible = true
                ivCategoryPlay.setImageResource(R.drawable.playicon)
                return
            }

            when (state) {
                PLAYING -> {
                    lottieLoadingAnimation.isInvisible = true
                    ivCategoryPlay.isVisible = true
                    ivCategoryPlay.setImageResource(R.drawable.pauseicon)
                }

                PREPARING, BUFFERING -> {
                    lottieLoadingAnimation.isVisible = true
                    ivCategoryPlay.isInvisible = true
                }

                else -> {
                    lottieLoadingAnimation.isInvisible = true
                    ivCategoryPlay.isVisible = true
                    ivCategoryPlay.setImageResource(R.drawable.playicon)
                }
            }
        }

    private fun updateFavoriteIcon(station: RadioStation, holder: ItemViewViewHolder) {
        val resId =
            if (station.isFavorite) R.drawable.favorite_20px_filled else R.drawable.favorite_20px
        holder.binding.ivCategoryFavourite.setImageResource(resId)
    }

    private inner class PlayOnclickListener(private val radioStation: RadioStation) :
        View.OnClickListener {
        override fun onClick(view: View) {
            radioStationActionHandler.onStationSelected(radioStation)
        }
    }

    /**
     * An inner class that implements the {@link View.OnClickListener} interface.
     * It is used to handle click events on favourite icon that represents a radio station's favorite status.
     * When the icon is clicked, this listener toggles the favorite status of the associated
     * {@link RadioStation} and notifies the {@link RadioStationActionHandler}.
     */
    private inner class FavouriteListener(
        private val radioStation: RadioStation,
    ) : View.OnClickListener {
        override fun onClick(view: View) {
            val isFavorite = radioStation.isFavorite

            radioStationActionHandler.apply {
                onToggleFavorite(radioStation, !isFavorite)
            }

        }
    }

    override fun getItemCount(): Int = radioStationItems.size

    inner class ItemViewViewHolder(val binding: CategoryradioitemBinding) :
        RecyclerView.ViewHolder(binding.root)

    fun updateStations(newStations: List<RadioStation>) {
        val diffCallback = StationDiffUtilCallback(radioStationItems, newStations)
        val diffResult = DiffUtil.calculateDiff(diffCallback)

        radioStationItems.clear()
        radioStationItems.addAll(newStations)

        diffResult.dispatchUpdatesTo(this)
    }


    fun setSelectedStationWithState(station: RadioStation, state: String) {
        this.state = state

        radioStationItems.forEach { it.isPlaying = false }

        val index = getPositionOfStationById(station.id)
        if (index != RecyclerView.NO_POSITION) radioStationItems[index].isPlaying = true

        notifyDataSetChanged()
    }

    fun updateFavorites(favorites: List<RadioStation>) {
        radioStationItems.forEach { station ->
            station.isFavorite = favorites.any { it.id == station.id }
        }
        notifyDataSetChanged()
    }

    private fun getPositionOfStationById(id: Int): Int {
        return radioStationItems.indexOfFirst { it.id == id }.takeIf { it >= 0 }
            ?: RecyclerView.NO_POSITION
    }

}
